File: dp_labs_and_solutions/dp_labs_and_solutions/dp_java_labs_and_solutions/command/memento_solution/CompoundCommand.java
// CompoundCommand.java

public class CompoundCommand extends Command
{
	public CompoundCommand()
	{
		myCommands = new DpVector<Command>();
	}

	public void add(Command aCommand)
	{
		if (aCommand == null)
		{
			throw new DpException("void CompoundCommand.add(Command aCommand) -- aCommand == null.");
		}
		myCommands.add(aCommand);
	}

	public void execute()
	{
		DpIterator<Command> aIterator = myCommands.makeIterator();
		while (!aIterator.isDone())
		{
			Command aCommand = aIterator.getCurrent();
			aCommand.execute();
			aIterator.advance();
		}
	}

	public void undo()
	{
		DpIterator<Command> aIterator = myCommands.makeReverseIterator();
		while (!aIterator.isDone())
		{
			Command aCommand = aIterator.getCurrent();
			aCommand.undo();
			aIterator.advance();
		}
	}

	private DpVector<Command> myCommands;
}

File: dp_labs_and_solutions/dp_labs_and_solutions/dp_cpp_labs_and_solutions/command/solution/compound_command.cpp
// compound_command.cpp

#include "compound_command.h"
#include "dp_exception.h"
#include "dp_iterator.h"
#include "dp_vector.h"

compound_command::compound_command()
{
}

void compound_command::add(command * a_command)
{
	if (a_command == nullptr)
	{
		throw dp_exception("void compound_command::add(command * a_command) -- a_command == nullptr.");
	}
	my_commands.add(a_command);
}

void compound_command::execute()
{
	dp_iterator<command *> * a_iterator = my_commands.make_iterator();
	while (!a_iterator->is_done())
	{
		command * a_command = a_iterator->get_current();
		a_command->execute();
		a_iterator->advance();
	}
}

void compound_command::undo()
{
	dp_iterator<command *> * a_iterator = my_commands.make_reverse_iterator();
	while (!a_iterator->is_done())
	{
		command * a_command = a_iterator->get_current();
		a_command->undo();
		a_iterator->advance();
	}
}
